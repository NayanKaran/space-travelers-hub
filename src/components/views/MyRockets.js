import React from 'react';
import { useSelector } from 'react-redux';
import ListGroup from 'react-bootstrap/ListGroup';

const MyRockets = () => {
  const state = useSelector((state) => state.entities.rockets);
  const rockets = state.filter((rocket) => rocket.reserved !== false);
  return (
    <div>
      <h2>My Rockets</h2>
      <ListGroup>
        {
          rockets.map((rocket) => (
            <ListGroup.Item key={rocket.id}>{rocket.name}</ListGroup.Item>
          ))
        }
      </ListGroup>
    </div>
  );
};

export default MyRockets;
