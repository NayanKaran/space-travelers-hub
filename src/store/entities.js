import { combineReducers } from '@reduxjs/toolkit';
import rocketsReducer from './rockets';
import missionsReducer from './missions';

export default combineReducers({
  rockets: rocketsReducer,
  missions: missionsReducer,
});
