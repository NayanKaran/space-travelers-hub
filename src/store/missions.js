import axios from 'axios';

// Actions
const FETCH_DATA = 'FETCH_DATA_FROM_API';
const TOGGLE_RESERVATION = 'TOGGLE_RESERVATION';

const missions = [];

// reducer
export default function missionsReducer(state = missions, action) {
  switch (action.type) {
    case FETCH_DATA: {
      return action.payload;
    }

    case TOGGLE_RESERVATION: {
      return state.map((mission) => {
        if (mission.mission_id !== action.id) {
          return mission;
        }
        return { ...mission, reserved: !mission.reserved };
      });
    }

    default:
      return state;
  }
}

// Action Creators
const fetchMission = (payload) => ({
  type: FETCH_DATA,
  payload,
});

export const joinMission = (id) => ({
  type: TOGGLE_RESERVATION,
  id,
});

/* eslint-disable camelcase */

// mission api
const url = 'https://api.spacexdata.com/v3/missions';

export const fetchData = () => async (dispatch) => {
  const response = await axios.get(url);
  const data = Object.entries(response.data).map((item) => {
    const {
      mission_id, mission_name, description,
    } = item[1];
    return {
      mission_id,
      mission_name,
      description,
      reserved: false,
    };
  });
  dispatch(fetchMission(data));
};
