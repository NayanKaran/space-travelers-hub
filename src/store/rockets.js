import { createSlice } from '@reduxjs/toolkit';

const slice = createSlice({
  name: 'rockets',
  initialState: [],
  reducers: {
    checkStatus: () => 'Under construction',
  },
});

export const { checkStatus } = slice.actions;

export default slice.reducer;
