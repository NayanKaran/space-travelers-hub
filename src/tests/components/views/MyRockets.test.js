import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import MyRockets from '../../../components/views/MyRockets';
import store from '../../../store/configureStore';

describe('Testing MyRockets component', () => {
  it('Testing with snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <MyRockets />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
