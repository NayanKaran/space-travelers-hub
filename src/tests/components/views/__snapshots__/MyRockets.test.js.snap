// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing MyRockets component Testing with snapshot 1`] = `
<div>
  <h2>
    My Rockets
  </h2>
  <div
    className="list-group"
    onKeyDown={[Function]}
  />
</div>
`;
